From fd887d16e6f7f71736432ec9f46d71e0c0211cca Mon Sep 17 00:00:00 2001
From: Marius Vlad <marius.vlad@collabora.com>
Date: Fri, 5 Mar 2021 21:58:40 +0200
Subject: [PATCH 7/8] kiosk-shell: Embed keyboard focus activation code

Just like desktop-shell, we shouldn't be dependent on having a keyboard
to be present in order to activate a window/surface.

Upstream-Status: Pending MR [https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/570]
Signed-off-by: Marius Vlad <marius.vlad@collabora.com>
---
 kiosk-shell/kiosk-shell.c | 55 ++++++++++++++++++++++++---------------
 1 file changed, 34 insertions(+), 21 deletions(-)

diff --git a/kiosk-shell/kiosk-shell.c b/kiosk-shell/kiosk-shell.c
index c25d139e..c73c6268 100644
--- a/kiosk-shell/kiosk-shell.c
+++ b/kiosk-shell/kiosk-shell.c
@@ -63,6 +63,20 @@ get_kiosk_shell_seat(struct weston_seat *seat)
 			    struct kiosk_shell_seat, seat_destroy_listener);
 }
 
+static void
+kiosk_shell_surface_deactivate(struct kiosk_shell_surface *shsurf)
+{
+	if (--shsurf->focus_count == 0)
+		weston_desktop_surface_set_activated(shsurf->desktop_surface, false);
+}
+
+static void
+kiosk_shell_surface_activate(struct kiosk_shell_surface *shsurf)
+{
+	if (shsurf->focus_count++ == 0)
+		weston_desktop_surface_set_activated(shsurf->desktop_surface, true);
+}
+
 /*
  * kiosk_shell_surface
  */
@@ -319,26 +333,7 @@ kiosk_shell_surface_create(struct kiosk_shell *shell,
 static void
 kiosk_shell_seat_handle_keyboard_focus(struct wl_listener *listener, void *data)
 {
-	struct weston_keyboard *keyboard = data;
-	struct kiosk_shell_seat *shseat = get_kiosk_shell_seat(keyboard->seat);
-
-	if (shseat->focused_surface) {
-		struct kiosk_shell_surface *shsurf =
-			get_kiosk_shell_surface(shseat->focused_surface);
-		if (shsurf && --shsurf->focus_count == 0)
-			weston_desktop_surface_set_activated(shsurf->desktop_surface,
-							     false);
-	}
-
-	shseat->focused_surface = weston_surface_get_main_surface(keyboard->focus);
-
-	if (shseat->focused_surface) {
-		struct kiosk_shell_surface *shsurf =
-			get_kiosk_shell_surface(shseat->focused_surface);
-		if (shsurf && shsurf->focus_count++ == 0)
-			weston_desktop_surface_set_activated(shsurf->desktop_surface,
-							     true);
-	}
+	/* unused */
 }
 
 static void
@@ -674,8 +669,26 @@ desktop_surface_committed(struct weston_desktop_surface *desktop_surface,
 		shsurf->view->is_mapped = true;
 		surface->is_mapped = true;
 
-		wl_list_for_each(seat, &shsurf->shell->compositor->seat_list, link)
+		wl_list_for_each(seat, &shsurf->shell->compositor->seat_list, link) {
+			struct kiosk_shell_seat *kiosk_seat = get_kiosk_shell_seat(seat);
 			weston_view_activate(shsurf->view, seat, 0);
+
+			if (kiosk_seat->focused_surface) {
+				struct kiosk_shell_surface *current_focus =
+					get_kiosk_shell_surface(kiosk_seat->focused_surface);
+				if (current_focus)
+					kiosk_shell_surface_deactivate(current_focus);
+			}
+
+			kiosk_seat->focused_surface = surface;
+
+			if (kiosk_seat->focused_surface) {
+				if (shsurf)
+					kiosk_shell_surface_activate(shsurf);
+			}
+		}
+
+
 	}
 
 	if (!is_fullscreen && (sx != 0 || sy != 0)) {
-- 
2.30.1

